### PR TITLE
Ai

### PR DESCRIPTION
**Pull request by konnorbeard**
Basic AI package. Currently only works with manualTurn and moveTypeManual. 

How it works: User enters "AI" as one of the player names. When it's AI's turn, we copy the game (cleverly, can't just do a raw copy because object references will still exist, and we can't do a raw JSON create object because we have circular references). Then we iterate through all possible moves and grab the legal ones, then evaluate them on copies of the game object. We then pick the best move, based on a user-defined "scoreBoard" function that should live in their game object. 

Todo: 
- AI should look > 1 move into future, should prune 
- Add different difficulties 
- Make it work with placeToken game types
- Test array board type (i tried to generalize but there's probably some bugs in there) 
- Display error messages, fail better (ie. if user is using graph board or something we should fail cleaner, not even sure what would happen now, might blow up) 
- Add documentation/tutorial on how to use, what limitations are, how to extend 

@andrewjmeier cr

### CODE REVIEW

File: tabletop/core/game.js [hunk 1]
```diff
@@ -3,6 +3,11 @@ var ManualTurn = require("./manual_turn.js");
 var Component = require("./component.js");
 var inherits = require('util').inherits;
 var _ = require('lodash');
+var $ = require('jquery');
```
Comment (andrewjmeier) [inline tabletop/core/game.js]: are we using jQuery here?

File: tabletop/core/game.js [hunk 2]
```diff
@@ -161,6 +166,15 @@ Game.prototype.nextPlayer = function() {
 };
 
 /**
+ * Returns the next player, but does not switch
+ * Override to provide more logic on determining the next player
+ * @returns {void}
```
Comment (andrewjmeier) [inline tabletop/core/game.js]: not returning void

Comment (konnorbeard) [inline tabletop/core/game.js]: yeah i gotta go through and add this JSdoc stuff for the other methods too. good catch here

File: tabletop/core/game.js [hunk 3]
```diff
@@ -293,4 +307,186 @@ Game.prototype.destroyToken = function(token) {
...
+        tilePos: tilePos, 
+        destPos: destPos
+      });
+    }, this);
+  }
+  
+  var bestResult = results[0];
+  var bestScore = 0;
+  
+  results.forEach(function(result) {
```
Comment (andrewjmeier) [inline tabletop/core/game.js]: could probably use lodash's find method and clean this up

File: tabletop/core/game.js (hunk 1 above)
Comment (konnorbeard) [inline tabletop/core/game.js]: woops i was but i removed it.

File: tabletop/core/game.js [hunk 4]
```diff
@@ -293,4 +307,186 @@ Game.prototype.destroyToken = function(token) {
...
+  
+  return legalMoves;
+};
+
+// Create copy of game and manually copy over 
+// tiles/tokens to remove references
+Game.prototype.copyGameStatus = function(game) {
+
+  //var newGame = $.extend(true, {}, game);
+  var newGame = Object.create(this);
```
Comment (andrewjmeier) [inline tabletop/core/game.js]: Can we use the same syntax that we use to create an object everywhere else?

File: tabletop/core/game.js [hunk 5]
```diff
@@ -293,4 +307,186 @@ Game.prototype.destroyToken = function(token) {
...
+  newGame.board = Object.create(this.board);
+  newGame.board.tokens = [];
+  newGame.board.tiles = [];
+  newGame.players = [];
+  var i, j = 0;
+  
+  // copy players, clear token arrays
+  for (i = 0; i < game.players.length; i++) { 
+    newGame.players.push(new Player());
+    Object.assign(newGame.players[i], game.players[i]);
```
Comment (andrewjmeier) [inline tabletop/core/game.js]: What is this doing?

File: tabletop/core/manual_turn.js [hunk 6]
```diff
@@ -58,13 +58,25 @@ function ManualTurn(game, startView, view, gameOverView, nextPlayerView) {
       waitingForMove: { 
         _onEnter: function() { 
           view.drawView();
-          console.log(this.game.getCurrentPlayer().name + ": Make your move.");
+          if (this.game.getCurrentPlayer().name == "AI") {
```
Comment (andrewjmeier) [inline tabletop/core/manual_turn.js]: Can we make a class for AIPlayer?

File: tabletop/core/manual_turn.js [hunk 7]
```diff
@@ -58,13 +58,25 @@ function ManualTurn(game, startView, view, gameOverView, nextPlayerView) {
...
+          }
         },
         
         makeMove : function() { 
           if (game.hasValidMove()) { 
-            game.executeMove();
-            this.transition("postTurn");
+            if (this.game.getCurrentPlayer().name == "AI") { 
+              var turnMap = this;
+              setTimeout(function() { game.executeMove(); turnMap.transition("postTurn"); }, 500);
```
Comment (andrewjmeier) [inline tabletop/core/manual_turn.js]: do we really want to add a fake delay?

**Comment by andrewjmeier** [comment]
Tried to play and it crashed picking a move about half way through the game

**Comment by andrewjmeier** [comment]
Looks pretty good though

**Comment by andrewjmeier** [comment]
I'm wondering if there's a place in the framework for a "move" class or something like that. The code basically only works for times when a move is from one tile to another involving a token. Could we instead have it function in a way where it didn't know/care what the move was? But just looped through the moves, executed the move, got the score, returned the best move?

File: tabletop/core/game.js (hunk 3 above)
Comment (konnorbeard) [inline tabletop/core/game.js]: yeah

File: tabletop/core/game.js (hunk 5 above)
Comment (konnorbeard) [inline tabletop/core/game.js]: copying the attributes from the real players to the copy

File: tabletop/core/manual_turn.js (hunk 6 above)
Comment (konnorbeard) [inline tabletop/core/manual_turn.js]: yeah probably good cal

File: tabletop/core/manual_turn.js (hunk 7 above)
Comment (konnorbeard) [inline tabletop/core/manual_turn.js]: for right now yeah. ideally we time analysis and if it's under .5 seconds then we add a delay. without a delay the move is made so quick it's hard to understand what happened

**Comment by konnorbeard** [comment]
weird. do you know what the error was when it crashed? or was it just looping? 

maybe if it's not manualMove or placeToken then we just have the user hand us a list of proposed moves. that could work nicely.

**Comment by andrewjmeier** [comment]
I think that the function that determines the moves should be passed the possible moves either way. Then you can have your function to automatically generate moves for some games but that way it's not doing both

File: src/checkers/checkers_game.js [hunk 8]
```diff
@@ -117,4 +114,14 @@ CheckersGame.prototype.playerDidWin = function(player) {
 };
 
 
+
+// takes in a player 
+// returns the score for the board based on the player passed in (ie. if the player 
+// has won it should return 10, if he loses should return -10) 
+CheckersGame.prototype.scoreBoard = function(player) {
```
Comment (andrewjmeier) [inline src/checkers/checkers_game.js]: This reads like scoreboard (as in the thing with numbers on it) not score as the verb. Could we use something like `evaluateState` or `evaluateBoard`

**Comment by andrewjmeier** [comment]
Let me know when you make the updates and want to try and merge

File: tabletop/core/game.js (hunk 3 above)
Comment (konnorbeard) [inline tabletop/core/game.js]: @andrewjmeier can lodash help here? i don't need the max object, i need the max object.score

File: tabletop/core/game.js (hunk 4 above)
Comment (konnorbeard) [inline tabletop/core/game.js]: @andrewjmeier Object.create() here copies the attributes of the game over to newGame

File: tabletop/core/game.js (hunk 3 above)
Comment (andrewjmeier) [inline tabletop/core/game.js]: Look at maxBy https://lodash.com/docs#maxBy

**Pull request closed without merging**